Clean up named tensor `propagate_names` API

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29322 named tensor support for torch.equal
* #29307 Make TensorName::unifyFromRight in-place for efficiency
* **#29239 Clean up named tensor `propagate_names` API**

There were a few main changes, summarized below.

Rename `propagate_names`
----------------------------------------------

There are two main APIs now, `propagate_names_if_nonempty(Tensor&,
ArrayRef<Dimname>)` and `propagate_names(Tensor&, ArrayRef<Dimname>)`

The former propagates names if they are not empty and the latter
unconditionally tries to propagate names.

`names` can be empty if name inference did not occur (see the next
section).

Removed usages of `optional` in name inference
----------------------------------------------

Previously, we used `optional<ArrayRef<Dimname>>` and
`optional<vector<Dimname>>`. `nullopt` represens that no name inference
happened.

The problem with this is that these types are not implicitly convertible
to each other and dealing with them is painful as a result (users have
to manually unwrap `optional<vector>` and convert to
`optional<arrayref>`.

To fix this, I rewrote most named inference functions to use an empty array as an
indicator value:
- If an array is empty, then no name inference occured
- If an array is not empty, then name inference occured.

Removed `vector<Dimname>&&` overloads
----------------------------------------------

These were originally meant for efficiency: instead of copying a vector
of names we could move it directly inside the tensor and replace the old
names. However, looking around the code base, we do copies for
`IntArrayRef` for sizes and strides instead of optimizing them, so the
perf gain is probably not critical. I removed `vector<Dimname>&&` overloads
to stop optimizing prematurely.

Furthermore, one potential design for a faster named inference api is
to construct names directly on a tensor's names object; in this design
there is also no `vector<Dimname>&&` overload.

Plans
----------------------------------------------

After this PR I'll keep attempting to cleaning up `propagate_names`
functions. There are a lot of `propagate_names_for_{blah}` functions
that exist that probably don't need to.

Test Plan:
- `python test/test_namedtensor.py -v`

Differential Revision: [D18350090](https://our.internmc.facebook.com/intern/diff/D18350090)

zou3519 committed:
Clean up named tensor `propagate_names` API

Previously, there were 8 `propagate_names` overloads. Now, there are 4.
There were two main changes, summarized below.

Removed most usages of `optional`
==================================

Previously, we used `optional<ArrayRef<Dimname>>` and
`optional<vector<Dimname>>`. `nullopt` represens that no name inference
happened.

The problem with this is that these types are not implicitly convertible
to each other and dealing with them is painful as a result (users have
to manually unwrap `optional<vector>` and convert to
`optional<arrayref>`.

To fix this, I rewrote most named inference functions to use an empty array as an
indicator value:
- If an array is empty, then no name inference occured
- If an array is not empty, then name inference occured.

Removed `vector<Dimname>&&` overloads
=====================================

These were originally meant for efficiency: instead of copying a vector
of names we could move it directly inside the tensor and replace the old
names. However, looking around the code base, we do copies for
`IntArrayRef` for sizes and strides instead of optimizing them, so the
perf gain is probably not critical. I removed `vector<Dimname>&&` overloads
to stop optimizing prematurely.

Furthoremore, one potential design for a faster named inference api is
to construct names directly on a tensor's names object; in this design
there is also no `vector<Dimname>&&` overload.

Plans
======

After this PR I'll keep attempting on cleaning up `propagate_names`
functions. There are a lot of `propagate_names_for_{blah}` functions
that exist that probably don't need to.

Test Plan:
- `python test/test_namedtensor.py -v`

[ghstack-poisoned]

zou3519 committed:
Update on "Clean up named tensor `propagate_names` API"


Previously, there were 8 `propagate_names` overloads. Now, there are 4.
There were two main changes, summarized below.

Removed most usages of `optional`
==================================

Previously, we used `optional<ArrayRef<Dimname>>` and
`optional<vector<Dimname>>`. `nullopt` represens that no name inference
happened.

The problem with this is that these types are not implicitly convertible
to each other and dealing with them is painful as a result (users have
to manually unwrap `optional<vector>` and convert to
`optional<arrayref>`.

To fix this, I rewrote most named inference functions to use an empty array as an
indicator value:
- If an array is empty, then no name inference occured
- If an array is not empty, then name inference occured.

Removed `vector<Dimname>&&` overloads
=====================================

These were originally meant for efficiency: instead of copying a vector
of names we could move it directly inside the tensor and replace the old
names. However, looking around the code base, we do copies for
`IntArrayRef` for sizes and strides instead of optimizing them, so the
perf gain is probably not critical. I removed `vector<Dimname>&&` overloads
to stop optimizing prematurely.

Furthermore, one potential design for a faster named inference api is
to construct names directly on a tensor's names object; in this design
there is also no `vector<Dimname>&&` overload.

Plans
======

After this PR I'll keep attempting to cleaning up `propagate_names`
functions. There are a lot of `propagate_names_for_{blah}` functions
that exist that probably don't need to.

Test Plan:
- `python test/test_namedtensor.py -v`

[ghstack-poisoned]

zou3519 committed:
Update on "Clean up named tensor `propagate_names` API"


There were a few main changes, summarized below.

Rename `propagate_names`
----------------------------------------------

There are two main APIs now, `propagate_names_if_nonempty(Tensor&,
ArrayRef<Dimname>)` and `propagate_names(Tensor&, ArrayRef<Dimname>)`

The former propagates names if they are not empty and the latter
unconditionally tries to propagate names.

`names` can be empty if name inference did not occur (see the next
section).

Removed usages of `optional` in name inference
----------------------------------------------

Previously, we used `optional<ArrayRef<Dimname>>` and
`optional<vector<Dimname>>`. `nullopt` represens that no name inference
happened.

The problem with this is that these types are not implicitly convertible
to each other and dealing with them is painful as a result (users have
to manually unwrap `optional<vector>` and convert to
`optional<arrayref>`.

To fix this, I rewrote most named inference functions to use an empty array as an
indicator value:
- If an array is empty, then no name inference occured
- If an array is not empty, then name inference occured.

Removed `vector<Dimname>&&` overloads
----------------------------------------------

These were originally meant for efficiency: instead of copying a vector
of names we could move it directly inside the tensor and replace the old
names. However, looking around the code base, we do copies for
`IntArrayRef` for sizes and strides instead of optimizing them, so the
perf gain is probably not critical. I removed `vector<Dimname>&&` overloads
to stop optimizing prematurely.

Furthermore, one potential design for a faster named inference api is
to construct names directly on a tensor's names object; in this design
there is also no `vector<Dimname>&&` overload.

Plans
----------------------------------------------

After this PR I'll keep attempting to cleaning up `propagate_names`
functions. There are a lot of `propagate_names_for_{blah}` functions
that exist that probably don't need to.

Test Plan:
- `python test/test_namedtensor.py -v`

[ghstack-poisoned]

zou3519 committed:
Update on "Clean up named tensor `propagate_names` API"


There were a few main changes, summarized below.

Rename `propagate_names`
----------------------------------------------

There are two main APIs now, `propagate_names_if_nonempty(Tensor&,
ArrayRef<Dimname>)` and `propagate_names(Tensor&, ArrayRef<Dimname>)`

The former propagates names if they are not empty and the latter
unconditionally tries to propagate names.

`names` can be empty if name inference did not occur (see the next
section).

Removed usages of `optional` in name inference
----------------------------------------------

Previously, we used `optional<ArrayRef<Dimname>>` and
`optional<vector<Dimname>>`. `nullopt` represens that no name inference
happened.

The problem with this is that these types are not implicitly convertible
to each other and dealing with them is painful as a result (users have
to manually unwrap `optional<vector>` and convert to
`optional<arrayref>`.

To fix this, I rewrote most named inference functions to use an empty array as an
indicator value:
- If an array is empty, then no name inference occured
- If an array is not empty, then name inference occured.

Removed `vector<Dimname>&&` overloads
----------------------------------------------

These were originally meant for efficiency: instead of copying a vector
of names we could move it directly inside the tensor and replace the old
names. However, looking around the code base, we do copies for
`IntArrayRef` for sizes and strides instead of optimizing them, so the
perf gain is probably not critical. I removed `vector<Dimname>&&` overloads
to stop optimizing prematurely.

Furthermore, one potential design for a faster named inference api is
to construct names directly on a tensor's names object; in this design
there is also no `vector<Dimname>&&` overload.

Plans
----------------------------------------------

After this PR I'll keep attempting to cleaning up `propagate_names`
functions. There are a lot of `propagate_names_for_{blah}` functions
that exist that probably don't need to.

Test Plan:
- `python test/test_namedtensor.py -v`

[ghstack-poisoned]

nairbv commented on aten/src/ATen/NamedTensorUtils.cpp:
```diff
@@ -128,51 +127,41 @@ static void assert_names_equal(DimnameList a, DimnameList b) {
-void propagate_names(Tensor& result, std::vector<Dimname>&& names, bool validate_names) {
-  propagate_names(result.unsafeGetTensorImpl(), std::move(names), validate_names);
+TensorImpl* propagate_names(TensorImpl* result, DimnameList names, bool validate_names) {
+  if (result->sizes().size() > 0) {
```
I think `result->dim()` is clearer if I'm correctly understanding this

zou3519 replied:
Yes, thanks for catching that

nairbv commented on aten/src/ATen/NamedTensorUtils.cpp:
```diff
@@ -461,20 +453,20 @@ optional<std::vector<Dimname>> compute_cat_outnames(TensorList tensors) {
 }
 
-optional<std::vector<Dimname>> compute_matmul_outnames(
+std::vector<Dimname> compute_matmul_outnames(
```
I like this change. I think I mentioned before I generally prefer avoiding optional/nullable lists wherever possible.

zou3519 replied:
I like the type safety that optional gives but I got too many headaches dealing with `optional<ArrayRef>` and `optional<vector>` so I decided to kill it here.

nairbv commented on aten/src/ATen/NamedTensorUtils.h:
```diff
@@ -43,26 +43,97 @@ unify_from_right(DimnameList names, DimnameList other, const char* action = "bro
-void propagate_names(TensorImpl* result, optional<std::vector<Dimname>>&& maybe_names, bool validate_names);
+// [NOTE] Writing name inference rules
+//
+// Operators that support named tensors generally look like the following:
```
I assume that in some cases ops that just call other ops could support named tensors without doing anything?

zou3519 replied:
Yes. I'll include that in

nairbv commented on aten/src/ATen/NamedTensorUtils.h:
```diff
@@ -43,26 +43,97 @@ unify_from_right(DimnameList names, DimnameList other, const char* action = "bro
+// pay no perf cost for it.
+
+// Propagates `names` to `result` if `names` is not empty.
+// `names` can be empty; see [NOTE] Writing name inference rules
```
this `see [NOTE]` seems recursive

zou3519 replied:
There's a space between the [NOTE] and this block that mentions the note. This block is supposed to be documentation for `propagate_names_if_nonempty`.

zou3519 replied:
I'll add an extra space to make it clearer

nairbv replied:
ah I see. some other ideas: you could start/end the note with a line of `////////////`'s or move the note above the start of the namespace, or use /**  */'s to better distinguish it from comments on the specific function below.

zou3519 committed:
Update on "Clean up named tensor `propagate_names` API"


There were a few main changes, summarized below.

Rename `propagate_names`
----------------------------------------------

There are two main APIs now, `propagate_names_if_nonempty(Tensor&,
ArrayRef<Dimname>)` and `propagate_names(Tensor&, ArrayRef<Dimname>)`

The former propagates names if they are not empty and the latter
unconditionally tries to propagate names.

`names` can be empty if name inference did not occur (see the next
section).

Removed usages of `optional` in name inference
----------------------------------------------

Previously, we used `optional<ArrayRef<Dimname>>` and
`optional<vector<Dimname>>`. `nullopt` represens that no name inference
happened.

The problem with this is that these types are not implicitly convertible
to each other and dealing with them is painful as a result (users have
to manually unwrap `optional<vector>` and convert to
`optional<arrayref>`.

To fix this, I rewrote most named inference functions to use an empty array as an
indicator value:
- If an array is empty, then no name inference occured
- If an array is not empty, then name inference occured.

Removed `vector<Dimname>&&` overloads
----------------------------------------------

These were originally meant for efficiency: instead of copying a vector
of names we could move it directly inside the tensor and replace the old
names. However, looking around the code base, we do copies for
`IntArrayRef` for sizes and strides instead of optimizing them, so the
perf gain is probably not critical. I removed `vector<Dimname>&&` overloads
to stop optimizing prematurely.

Furthermore, one potential design for a faster named inference api is
to construct names directly on a tensor's names object; in this design
there is also no `vector<Dimname>&&` overload.

Plans
----------------------------------------------

After this PR I'll keep attempting to cleaning up `propagate_names`
functions. There are a lot of `propagate_names_for_{blah}` functions
that exist that probably don't need to.

Test Plan:
- `python test/test_namedtensor.py -v`

[ghstack-poisoned]

zou3519 committed:
Update on "Clean up named tensor `propagate_names` API"


There were a few main changes, summarized below.

Rename `propagate_names`
----------------------------------------------

There are two main APIs now, `propagate_names_if_nonempty(Tensor&,
ArrayRef<Dimname>)` and `propagate_names(Tensor&, ArrayRef<Dimname>)`

The former propagates names if they are not empty and the latter
unconditionally tries to propagate names.

`names` can be empty if name inference did not occur (see the next
section).

Removed usages of `optional` in name inference
----------------------------------------------

Previously, we used `optional<ArrayRef<Dimname>>` and
`optional<vector<Dimname>>`. `nullopt` represens that no name inference
happened.

The problem with this is that these types are not implicitly convertible
to each other and dealing with them is painful as a result (users have
to manually unwrap `optional<vector>` and convert to
`optional<arrayref>`.

To fix this, I rewrote most named inference functions to use an empty array as an
indicator value:
- If an array is empty, then no name inference occured
- If an array is not empty, then name inference occured.

Removed `vector<Dimname>&&` overloads
----------------------------------------------

These were originally meant for efficiency: instead of copying a vector
of names we could move it directly inside the tensor and replace the old
names. However, looking around the code base, we do copies for
`IntArrayRef` for sizes and strides instead of optimizing them, so the
perf gain is probably not critical. I removed `vector<Dimname>&&` overloads
to stop optimizing prematurely.

Furthermore, one potential design for a faster named inference api is
to construct names directly on a tensor's names object; in this design
there is also no `vector<Dimname>&&` overload.

Plans
----------------------------------------------

After this PR I'll keep attempting to cleaning up `propagate_names`
functions. There are a lot of `propagate_names_for_{blah}` functions
that exist that probably don't need to.

Test Plan:
- `python test/test_namedtensor.py -v`

Differential Revision: [D18350090](https://our.internmc.facebook.com/intern/diff/D18350090)

[ghstack-poisoned]
